cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java

```diff
@@ -0,0 +1,271 @@
+    }
+    job.getConfiguration().set("input.path.name", config.path);
+    if (config.timeTable != null) {
+      table = context.getDataset(config.timeTable);
+    }
+
+    FileInputFormat.setInputPathFilter(job, S3Filter.class);
+    FileInputFormat.addInputPath(job, new Path(config.path));
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Text> input, Emitter<StructuredRecord> emitter) throws Exception {
+    StructuredRecord output = StructuredRecord.builder(DEFAULT_SCHEMA)
+      .set("ts", System.currentTimeMillis())
+      .set("body", input.getValue().toString())
+      .build();
+    emitter.emit(output);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    if (succeeded && table != null && config.fileRegex != null && config.fileRegex.equals("timefilter")) {
+      table.write("lastTimeRead", currentTime);
+    }
+  }
+
+  /**
+   * Filter class to filter out filenames in the input path.
+   */
+  public static class S3Filter extends Configured implements PathFilter {
+    private boolean useTimeFilter;
+    private Pattern regex;
+    private String pathName;
+
+    @Override
+    public boolean accept(Path path) {
+      String filename = path.toString();
+      System.out.println("attempt :" + filename);
+      //InputPathFilter will first check the directory if a directory is given
+      if (filename.equals(pathName) || filename.equals(pathName + "/")) {
+        return true;
+      }
+
+      if (useTimeFilter) {
+        //use stateful time filter
+        if (table != null) {
+          String lastRead = Bytes.toString(table.read("lastTimeRead"));
+          SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd-HH-mm");
+          Date dateLastRead;
+          try {
+            dateLastRead = sdf.parse(lastRead);
+          } catch (Exception e) {
+            dateLastRead = new Date();
+            dateLastRead.setTime(0);
+          }
+
+          Date fileDate;
+          try {
+            System.out.println("FILENAME puppies: " + path.toString());
+            System.out.println("PATHNAME :" + pathName);
+            fileDate = sdf.parse(path.getName().substring(0, DATE_LENGTH));
+          } catch (ParseException pe) {
+            //this should never happen
+            LOG.warn("Couldn't parse file: " + path.getName());
+            fileDate = prevMinute;
+          }
+
+          if (fileDate.compareTo(dateLastRead) > 0 && fileDate.compareTo(prevMinute) <= 0) {
+            return true;
+          }
+          return false;
+        } else {
+          //use hourly time filter
+          Date prevHour = new Date();
+          prevHour.setTime(System.currentTimeMillis() - TimeUnit.HOURS.toMillis(1));
+          String currentTime = new SimpleDateFormat("yyyy-MM-dd-HH").format(prevHour);
+
+          if (filename.contains(currentTime)) {
+            return true;
+          }
+          return false;
+        }
+      }
+
+      //use regex
```

Comment:
This comment gives no value. Maybe better as "filter by file name using regex from configuration".